### PR TITLE
[HIP] fix host min/max in header

### PR DESCRIPTION
**Pull request by yxsamliu**
CUDA defines min/max functions for host in global namespace. HIP header needs to define them too to be compatible. Currently only min/max(int, int) is defined. This causes wrong result for arguments that are out of range for int. This patch defines host min/max functions to be compatible with CUDA.

Fixes: SWDEV-446564

### CODE REVIEW

File: clang/lib/Headers/__clang_hip_math.h [hunk 1]
```diff
@@ -1306,14 +1306,50 @@ float min(float __x, float __y) { return __builtin_fminf(__x, __y); }
...
+  -> typename std::remove_reference<decltype(__a < __b ? __a : __b)>::type { \
+  return (__a < __b) ? __a : __b; \
+} \
+static inline auto max(const type1 __a, const type2 __b) \
+  -> typename std::remove_reference<decltype(__a > __b ? __a : __b)>::type { \
+  return (__a > __b) ? __a : __b; \
+}
+
+// Define min and max functions for same type comparisons
+DEFINE_MIN_MAX_FUNCTIONS(int, int)
```
Comment (jhuber6) [inline clang/lib/Headers/__clang_hip_math.h:1325]: Could we not do something like this w/ the appropriate static assertion? Or is there an important restriction on the specific types for this function.
```c
template <typename T, typename U>
static inline auto min(const T &__a, const U &__b) {
  return (__a < __b) ? __a : __b;
}
```

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h:1325]: it will return reference of `__a` or `__b`. that won't work since it is reference to stack var

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h:1325]: sorry I was wrong about the reference.

The issue with the template approach is that it has subtle differences regarding overloading resolution compared with CUDA, which could lead to incompatibility.

Comment (jhuber6) [reply clang/lib/Headers/__clang_hip_math.h]: Could we not do stuff like `static_assert` where we check if both types are the same modulo `unsigned` or `const`? I'm assuming that would be similar. I'd just prefer to avoid macro magic if possible, but it we really need it for compatibility reasons we can use it.

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h]: it is not just limiting types. template function has lower precedence than functions in overloading resolution. If users using std::min we will get different behavior than CUDA

File: clang/lib/Headers/__clang_hip_math.h [hunk 2]
```diff
@@ -1306,15 +1306,68 @@ float min(float __x, float __y) { return __builtin_fminf(__x, __y); }
...
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long long, unsigned long long,
+                         unsigned long long)
+
+// Define min and max functions for all mixed type comparisons
+DEFINE_MIN_MAX_FUNCTIONS(unsigned int, int, unsigned int)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned int, unsigned int, int)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long, long, unsigned long)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long, unsigned long, long)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long long, long long, unsigned long long)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long long, unsigned long long, long long)
```
Comment (Artem-B) [inline clang/lib/Headers/__clang_hip_math.h:1338]: I assume these are needed in order to avoid errors about ambiguous overload resolution when we pass signed/unsigned arguments.

Normally, if we were to use `std::min()` function, the user would have to explicitly cast arguments or use `std::min<unsigned>()` to resolve the issue.

Implicitly converting int->unsigned under the hood is probably not a good idea here as we do not know what the user needs/wants and whether it's a WAI or an error. For min/max converting a negative argument into an unsigned would probably be an error. I think we do need to force users to use one of the all-signed or all-unsigned variants here, too, same as with std::min/max.

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h:1338]: > I assume these are needed in order to avoid errors about ambiguous overload resolution when we pass signed/unsigned arguments.
> 
> Normally, if we were to use `std::min()` function, the user would have to explicitly cast arguments or use `std::min<unsigned>()` to resolve the issue.
> 
> Implicitly converting int->unsigned under the hood is probably not a good idea here as we do not know what the user needs/wants and whether it's a WAI or an error. For min/max converting a negative argument into an unsigned would probably be an error. I think we do need to force users to use one of the all-signed or all-unsigned variants here, too, same as with std::min/max.

You are right about that std::min/max does not allow mixed signed/unsigned arguments.

Unfortunately, that is how CUDA defines its host min/max functions, e.g. https://godbolt.org/z/Wxxq9dv91

Not following this could cause incompatibility and regressions.

Comment (Artem-B) [reply clang/lib/Headers/__clang_hip_math.h:1338]: Not everything CUDA does is the right model to follow. This may be one of the cases where we should improve things, if we can, instead of just copying the broken behavior. Not adding problematic things is easier than removing them later, when they are used, intentionally or not.

Considering that HIP currently does not have those functions, it would suggest that there is probably no existing HIP code depending on them. Existing cuda code which may need those functions will need some amount of porting to HIP, anyway, so fixing the source code could be done as part of the porting effort.

We could put those mixed min/max functions under some preprocessor guard, which would keep them disabled by default. If someone desperately needs them, they would have to specify `-DPLEASE_ENABLE_BROKEN_MINMAX_ON_MIXED_SIGNED_UNSIGNED_TYPES`.

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h:1342]: OK. will define them only if `__HIP_DEFINE_MIXED_HOST_MIN_MAX__` is defined.

File: clang/lib/Headers/__clang_hip_math.h [hunk 3]
```diff
@@ -1306,15 +1306,73 @@ float min(float __x, float __y) { return __builtin_fminf(__x, __y); }
...
+DEFINE_MIN_MAX_FUNCTIONS(long, long, long)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long, unsigned long, unsigned long)
+DEFINE_MIN_MAX_FUNCTIONS(long long, long long, long long)
+DEFINE_MIN_MAX_FUNCTIONS(unsigned long long, unsigned long long,
+                         unsigned long long)
+
+// CUDA defines host min/max functions with mixed signed/unsgined integer
+// parameters where signed integers are casted to unsigned integers. However,
+// this may not be users' intention. Therefore do not define them by default
+// unless users specify -D__HIP_DEFINE_MIXED_HOST_MIN_MAX__.
```
Comment (Artem-B) [inline clang/lib/Headers/__clang_hip_math.h:1335]: Nit: signed integers are implicitly promoted to unsigned ones due to the integer promotion rules. Cast would imply intentional cast and we're not doing that.

I'd rephrase it a bit along the lines of:

The routines below will perform unsigned comparison, which may produce invalid results if a signed integer was passed unintentionally. We do not want it happen silently, and do not provide these overloads by default. However for compatibility with CUDA, we allow them, if explicitly requested by the user by defining `__HIP_DEFINE_MIXED_HOST_MIN_MAX__`.

**Review by Artem-B — APPROVED** [review]

File: clang/lib/Headers/__clang_hip_math.h (hunk 3 above)
Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h:1335]: will do

Comment (yxsamliu) [reply clang/lib/Headers/__clang_hip_math.h]: I reverted this PR due to regression in hipCUB:

https://github.com/ROCm/hipCUB/blob/develop/hipcub/include/hipcub/backend/rocprim/device/device_spmv.hpp#L142

hipCUB/hipcub/include/hipcub/backend/rocprim/device/device_spmv.hpp:142:33: error: call to 'min' is ambiguous

it is a call of min(int, unsigned int). The code is likely ported from CUDA.

Probably I need to define those functions with mixed args by default to avoid regressions.

**Comment by Artem-B** [comment]
> Probably I need to define those functions with mixed args by default to avoid regressions.

Are there any other regressions? Can hupCUB be fixed intsead? While their use case is probably benign, I'd rather fix the user code, than propagate CUDA bugs into HIP.

**Comment by yxsamliu** [comment]
> > Probably I need to define those functions with mixed args by default to avoid regressions.
> 
> Are there any other regressions? Can hupCUB be fixed instead? While their use case is probably benign, I'd rather fix the user code, than propagate CUDA bugs into HIP.

So far we only found this issue in our internal CI. I will try asking hipCUB to fix it on their side.

**Comment by yxsamliu** [comment]
> > > Probably I need to define those functions with mixed args by default to avoid regressions.
> > 
> > 
> > Are there any other regressions? Can hupCUB be fixed instead? While their use case is probably benign, I'd rather fix the user code, than propagate CUDA bugs into HIP.
> 
> So far we only found this issue in our internal CI. I will try asking hipCUB to fix it on their side.

hipCUB issue opened https://github.com/ROCm/hipCUB/issues/343

**Comment by yxsamliu** [comment]
found another library using mixed min: https://github.com/ROCm/Tensile/issues/1977

**Comment by yxsamliu** [comment]
yet another usage of mixed signed/unsigned min https://github.com/ROCm/hipBLASLt/issues/1227
